pants_plugins/sendwave/pants_docker/package.py

```diff
@@ -223,16 +220,36 @@ async def package_into_image(
     process_path = docker_paths.first_path.path
     # build an list of arguments of the form ["-t",
     # "registry/name:tag"] to pass to the docker executable
+    full_target_name = _build_full_target_name(
+        target_name, field_set.registry.value
+    )
     tag_arguments = _build_tag_argument_list(
-        target_name, field_set.tags.value or [], field_set.registry.value
+        full_target_name, field_set.tags.value or []
     )
     # create the image
-    process_args = [process_path, "build"]
+    # We need to enable BuildKit if we plan to read and write to a remote cache
+    cli_command = (
+        ["buildx", "build"] if docker.options.cache_from else ["build"]
+    )
+    process_args = [process_path] + cli_command
     process_args.extend(tag_arguments)
     process_args.append(".")  # use current (sealed) directory as build context
     if docker.options.report_progress:
         process_args.append("--progress")
         process_args.append("plain")
+    if docker.options.cache_from:
+        registry_separator = (
+            "" if docker.options.cache_from.endswith("/") else "/"
+        )
+        cache_registry = (
+            f"{docker.options.cache_from}{registry_separator}{full_target_name}"
+        )
+        process_args += [
+            "--build-arg",
+            "BUILDKIT_INLINE_CACHE=1",
+            "--cache-from",
+            cache_registry,
+        ]
```

Comment:
I could see an argument for wanting to configure separate arguments for the "make this image a cache" flag (`--build-arg`) and the "cache from a remote image" flag (`--cache-from`), but I don't think our use-case requires that distinction, so since it would make configuration more complicated to require two arguments I decided to just go ahead and collapse them (meaning every build with `cache_from` configured will both pull the cache from the configured registry and add a cache manifest to the new image).

Reply:
that seems fine - It might be nice to pull this logic out into it's own function or something, for code hygiene!

Will this work the first time it's run? Since `--cache-from` needs to be set in order to build using the cache, but the image we specify won't be valid to use as the cache?

Reply:
I'm pretty sure it will work the first time -- when I tested this locally with a remote cache that didn't exist, it seemed like it just fell back to a local cache -- but I will double check!